### PR TITLE
Dragons drop devoured bodies, no stomach attacks and passive regeneration

### LINKED ISSUE #10816: Bodies will rot inside space dragon
## Description
Bodies will rot inside space dragon. Bodies becoming unsalvageable over time inside dragon that ate them is plausible, buzzing of flies following dragon - not very.

**Reproduction**
- Eat corpses as dragon
- Let some time pass
- Hear annoying buzzing of flies

**Additional context**
Give dragon `AntiRottingContainer`, to ensure that people can be rescued from cowardly dragon that went to hide in space for so long rot gibbing becomes a concern?

### LINKED ISSUE #16726: people still bleed inside dragons stomach
## Description
title

**Reproduction**
1. spawn dragon and urist
2. eat urist, hes probably bleeding by now
3. blood phases through you and onto the floor

### LINKED ISSUE #20405: Eaten Bodies Cannot Be Recovered From Space Dragon
## Description

Round 33913. A space dragon was butchered, leaving no bodies to recover. Players affected blackscreened and had to reconnect to see again.

**Reproduction**

Butcher a space dragon with a player's body inside.

### LINKED ISSUE #20935: Space Dragon's Devour destroy corpses/butcher doesn't give the corpses back
## Description

Whenever a Space Dragon devour a corpse and is cut open, no corpse appear 

**Reproduction**

1-Be a space dragon
2-Eat people
3-Die
4-Butcher the dragon
5-No corpses

**Screenshots**

**Additional context**

### LINKED ISSUE #21449: Should people be able to attack dragons from inside them?
image

This question was raised on admin help.
Whisper and I disagreed on wether or not this is intended.

I'm of the opinion that it's not intended, as it offers zero counterplay to the dragon player.
Opinions?

### LINKED ISSUE #23540: Devoured zombies can attack a Space Dragon from the inside
## Description
If you devour a zombie as a space dragon, they will eventually revive inside of you and be able to attack you. If this happens, you can't do anything to prevent it and WILL get forcefully zombified unless the person decides to be nice.

I know #21449 exists already, but this is a much more concrete example of why this shouldn't exist. It's not difficult to perform at all; if you just devour a zombie as a space dragon, you're fucked.

**Reproduction**
1. Be a space dragon
2. Devour a zombified human
3: Have the zombified humanbite your from inside
4: God fucking damn it now I'm a zombie and my carps are going to attack me as soon as they come out of the portal.

### PR DESCRIPTION
**Pull request by Arendian**
## About the PR

resolves #21449
resolves #23540
resolves #20935
resolves #20405
resolves #16726
resolves #10816

Adds a Devoured component to any devoured entities that passively deals caustic damage while an entity has it.
The damage is based on the entities MobState, with it being the highest while the entity is alive and the lowest while the entity is dead.
Damage dealt in this way is currently capped at 100 damage more than when an entity is considered dead.
Entities get the Pacified component while devoured.

Devoured bodies won't be deleted when the dragon is gibbed, butchered or when the dragon despawns from not making enough rifts.
After being devoured entities will stop bleeding, this is to prevent the dragon leaving a trail of blood puddles behind them.

Devoured entities won't rot, not too sure about this one because it does remove a lot of the "danger feeling" a dragon has if everyone can just have a happy ending the moment the dragon is killed.

Dragons now have a low amount of passive healing, and  in exchange no ichor gets injected anymore when they devour a humanoid entity.
Devouring a humanoid entity permanently increases their passive healing.

## Why / Balance

Entities in a dragon's stomach should take damage from the stomach acids it (probably?) has.
The damage is lowered on entities in a critical or dead state so that a timely retrieval of the body will need less healing or may even save them before they die.

A dragon can't defend itself against enemies that are inside their stomach, making the devoured entity pacified prevents this.

If the crew manages to kill the dragon they should be able to retrieve the devoured bodies for reviving/burial/borging.

Having the passive healing increase for every devoured entity will reward the dragon player if they manage to devour a lot of crew, and can develop them into an even more dangerous threat if the crew doesn't manage to stop the dragon early on.
 Some feedback on the amount of healing would be appreciated.

## Technical details

When a space dragon devours an entity, the entity will get the Devoured, Pacified and PassiveDamage component if they don't have these yet. These are removed again when the entities are released from the stomach.

The devoured component passively deals damage to an entity that has it, with a multiplier for this damage based on the entities MobState.

Before the dragon gets deleted the stomach will be emptied.

## Media

Video is outdated, no chemicals are injected anymore, but the current version still achieves the same result.

https://github.com/space-wizards/space-station-14/assets/137322659/fbf1309f-ecef-4a47-95c2-28643c755c9b

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Dygon
- add: Space dragons now have passive health regeneration that increases with every humanoid they devour.
- remove: Space dragons don't get an ichor injection anymore after devouring a humanoid.
- tweak: Bodies devoured by a space dragon don't rot.
- fix: Mobs devoured by a space dragon now won't be able to attack and will continuously take damage.
- fix: Bodies devoured by a space dragon won't be deleted anymore when the dragon is butchered.

### CODE REVIEW

File: Content.Server/Devour/DevourSystem.cs [hunk 1]
```diff
@@ -32,6 +44,9 @@ private void OnDoAfter(EntityUid uid, DevourerComponent component, DevourDoAfter
             if (component.ShouldStoreDevoured && args.Args.Target is not null)
             {
                 ContainerSystem.Insert(args.Args.Target.Value, component.Stomach);
+
+                //Stops the bleeding of the devoured target so the dragon doesn't leave a trail of random blood puddles behind
+                _bloodstreamSystem.TryModifyBleedAmount(args.Args.Target.Value, -100);
             }
```
Comment (metalgearsloth) [inline Content.Server/Devour/DevourSystem.cs:50]: This should be fixed on the bloodstream end and not worked around externally.

File: Content.Server/Devour/DevourSystem.cs [hunk 2]
```diff
@@ -45,5 +60,19 @@ private void OnDoAfter(EntityUid uid, DevourerComponent component, DevourDoAfter
...
     }
+
+    private void EmptyStomach(EntityUid uid, DevourerComponent component)
+    {
+        //Remove entities from the stomach before deletion.
+        foreach (var entity in component.Stomach.ContainedEntities.ToArray())
+        {
+            _containerSystem.Remove(entity, component.Stomach);
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Server/Devour/DevourSystem.cs:71]: Isn't this just EmptyContainer?

File: Content.Shared/Devour/Components/DevourerComponent.cs [hunk 3]
```diff
@@ -71,6 +72,18 @@ public sealed partial class DevourerComponent : Component
     [ViewVariables(VVAccess.ReadWrite), DataField("healRate")]
     public float HealRate = 15f;
 
+    /// <summary>
+    /// The chemical ID's injected into the victim upon being devoured.
+    /// </summary>
+    [DataField("stomachChemicals", customTypeSerializer: typeof(PrototypeIdListSerializer<ReagentPrototype>))]
+    public List<string> StomachChemicals = new();
```
Comment (metalgearsloth) [inline Content.Shared/Devour/Components/DevourerComponent.cs:79]: Please update to modern codebase standards i.e. no datafield string and using protoid.

File: Content.Shared/Devour/Components/DevourerComponent.cs [hunk 4]
```diff
@@ -71,6 +72,18 @@ public sealed partial class DevourerComponent : Component
...
+
+    /// <summary>
+    /// The amount of stomach chemicals  injected into the devoured entity.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("damageRate")]
+    public float DamageRate = 50f;
```
Comment (metalgearsloth) [inline Content.Shared/Devour/Components/DevourerComponent.cs:85]: DamageRate does not correspond to what this is actually doing as it has nothing to do with damage.

You're better off just using a solution instead.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by Kadeo64** [comment]
I'm against the whole injecting carpotoxin into people's bodies, I think pax is enough. The crit gasping makes it obvious the dragon has devoured someone and is currently healing, I think pax is all the dragon needs.

**Comment by Kadeo64** [comment]
also devoured entities not rotting is a good change, because it's annoying for people to rot when the dragon devours three people and wild goose chases for 15 minutes.

File: Content.Shared/Devour/Components/DevourerComponent.cs (hunk 3 above)
Comment (Arendian) [reply Content.Shared/Devour/Components/DevourerComponent.cs]: I'll do a reread of the new conventions before I will make any new PR's

File: Content.Server/Devour/DevourSystem.cs (hunk 2 above)
Comment (Arendian) [reply Content.Server/Devour/DevourSystem.cs]: Apparently, yes!

File: Content.Server/Devour/DevourSystem.cs (hunk 1 above)
Comment (Arendian) [reply Content.Server/Devour/DevourSystem.cs]: I'm not 100% sure I understood what you meant, does the event I made come close to it?

File: Content.Shared/Devour/Components/DevourerComponent.cs (hunk 4 above)
Comment (Arendian) [reply Content.Shared/Devour/Components/DevourerComponent.cs]: I tried following the naming already used in the component, the datafield keeping track of the amount of ichor injected was called HealRate.

**Comment by Arendian** [comment]
> edit: forgor zombies existed

The carpotoxin is mainly for the zombies yes, also fits the theme of the dragon since it's distant family of the space carps.

**Comment by Arendian** [comment]
I'm converting the injection parts to a system that adds pacifieComponent and passiveDamageComponent

**Comment by Arendian** [comment]
I added a system that applies pacified and a passive damage effect to any devoured entities. I also replaced the ichor injection with a passive health regeneration that increases with every humanoid devoured.

**Comment by Arendian** [comment]
The passive healing change might be out of scope for this PR so I can put it in another one if that makes this one easier to review,

**Comment by deltanedas** [comment]
> Devouring a humanoid entity permanently increases their passive healing.

juggernaut real

i like no rotting in stomach since once you open dragon up the rot timer begins and med has to run around like mad to try morgue/bag and heal everyone

File: Content.Server/Body/Systems/BloodstreamSystem.cs [hunk 5]
```diff
@@ -458,7 +458,7 @@ public void ChangeBloodReagent(EntityUid uid, string reagent, BloodstreamCompone
     /// <summary>
     ///     Stops the bleeding of the devoured target so the dragon doesn't leave a trail of random blood puddles behind
     /// </summary>
-    private void OnDevoured(EntityUid uid, BloodstreamComponent component, OnDevouredEvent args)
+    private void OnDevoured(EntityUid uid, BloodstreamComponent component, ref OnDevouredEvent args)
```
Comment (deltanedas) [inline Content.Server/Body/Systems/BloodstreamSystem.cs:461]: Entity<BloodstreamComponent> ent and use ent.Comp to get the component, do this for all stuff

File: Content.Server/Devour/DevouredSystem.cs [hunk 6]
```diff
@@ -24,42 +24,42 @@ public override void Initialize()
...
         {
-            SetStomachDamage(mobState.CurrentState, component, passiveDamage);
+            SetStomachDamage(mobState.CurrentState, entity.Comp, passiveDamage);
         }
 
         //Sets the max damage to the damage needed for the mob to be considered dead +
         //the damage cap specified in the component. This can be used to make reviving take more work if
         //the target has been devoured for a long time.
-        if (TryComp<MobThresholdsComponent>(uid, out var mobThresholds))
+        if (TryComp<MobThresholdsComponent>(entity.Owner, out var mobThresholds))
```
Comment (deltanedas) [inline Content.Server/Devour/DevouredSystem.cs:55]: Entity automatically casts to EntityUid so just pass entity to trycomp

File: Content.Server/Devour/DevouredSystem.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            {
+                passiveDamage.Damage -= component.StomachDamage * component.CurrentModifier;
+                passiveDamage.AllowedStates = component.OriginalAllowedMobStates;
+                passiveDamage.Damage.TrimZeros();
+            }
+        }
+
+        //Remove the pacified component if the target originally didn't have it.
+        if (!component.OriginallyPacified)
+            RemComp<PacifiedComponent>(uid);
```
Comment (metalgearsloth) [inline Content.Server/Devour/DevouredSystem.cs:98]: Why is this needed?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
Did a partial review

File: Content.Server/Devour/DevouredSystem.cs (hunk 7 above)
Comment (Arendian) [reply Content.Server/Devour/DevouredSystem.cs:98]: It's so people that had the pacified component before they got devoured(thieves and people with the pacifist trait) will still have the component. But I'll change it to actionblocker like you suggested.

File: Content.Server/Devour/Components/DevouredComponent.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    [DataField]
+    public bool OriginallyPacified;
+
+    /// <summary>
+    ///     Stores the original MobStates the entity was allowed to be healed in.
+    ///     This is so it can be returned to it's original state when the component is removed.
+    /// </summary>
+    [DataField]
+    public List<MobState>? OriginalAllowedMobStates;
+
```
[the remark this answers is not in the document]
Comment (Arendian) [reply Content.Server/Devour/Components/DevouredComponent.cs]: I fixed the prediction issue

File: Content.Server/Dragon/DragonSystem.cs [hunk 9]
```diff
@@ -98,7 +100,9 @@ public override void Update(float frameTime)
             if (comp.RiftAccumulator >= comp.RiftMaxAccumulator)
             {
                 Roar(uid, comp);
-                QueueDel(uid);
+                if(TryComp<DevourerComponent>(uid, out var devourComponent))
+                    _devour.EmptyStomach(uid, devourComponent);
+                _body.GibBody(uid);
```
Comment (AJCM-git) [inline Content.Server/Dragon/DragonSystem.cs:105]: wouldnt gibbing already empty the stomach since it would raise the event that gets handled by the devour system?

File: Content.Server/Devour/DevourSystem.cs [hunk 10]
```diff
@@ -1,49 +1,89 @@
...
+    /// </summary>
+    private void OnBeingGibbed(EntityUid uid, DevourerComponent component, BeingGibbedEvent args)
+    {
+        EmptyStomach(uid, component);
+    }
+
+    /// <summary>
+    /// Empties the stomach when component is removed.
+    /// </summary>
+    private void OnRemoved(EntityUid uid, DevourerComponent component, ComponentRemove args)
```
Comment (AJCM-git) [inline Content.Server/Devour/DevourSystem.cs:85]: i could be misremembering but i could swear gibbing raised ComponentRemove since it deletes the entity, is the other sub necessary?

File: Content.Server/Zombies/ZombieSystem.Transform.cs [hunk 11]
```diff
@@ -116,7 +116,13 @@ public void ZombifyEntity(EntityUid target, MobStateComponent? mobState = null)
             //This is needed for stupid entities that fuck up combat mode component
             //in an attempt to make an entity not attack. This is the easiest way to do it.
             var combat = EnsureComp<CombatModeComponent>(target);
-            RemComp<PacifiedComponent>(target);
+
+            //Only remove the pacified component if the target hasn't been devoured.
+            //This is to prevent entities that turn into a zombie while devoured
+            //from attacking the entity that devoured them from the inside.
+            if(!HasComp<DevouredComponent>(target))
+                RemComp<PacifiedComponent>(target);
```
Comment (AJCM-git) [inline Content.Server/Zombies/ZombieSystem.Transform.cs:124]: dont you cancel the attack attempt now?

File: Content.Shared/Devour/Components/DevouredComponent.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+[RegisterComponent, NetworkedComponent]
+[Access(typeof(DevouredSystem))]
+public sealed partial class DevouredComponent : Component
+{
+    /// <summary>
+    ///     Stores the original MobStates the entity was allowed to be healed in.
+    ///     This is so it can be returned to it's original state when the component is removed.
+    /// </summary>
+    [DataField]
+    public List<MobState>? OriginalAllowedMobStates;
```
Comment (AJCM-git) [inline Content.Shared/Devour/Components/DevouredComponent.cs:17]: I really dislike storing this, is it really necessary?

File: Content.Server/Devour/DevourSystem.cs [hunk 13]
```diff
@@ -1,49 +1,89 @@
...
         {
-            ichorInjection.ScaleSolution(0.5f);
-
-            if (component.ShouldStoreDevoured && args.Args.Target is not null)
+            if(TryComp<PassiveDamageComponent>(uid, out var passiveHealing))
             {
-                ContainerSystem.Insert(args.Args.Target.Value, component.Stomach);
+                if (component.PassiveDevourHealing != null)
+                {
+                    passiveHealing.Damage += component.PassiveDevourHealing;
```
Comment (AJCM-git) [inline Content.Server/Devour/DevourSystem.cs:46]: how many bodies do a dragon stomach fit?

File: Content.Shared/Devour/Components/DevouredComponent.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    ///     Damage multiplier if the devoured entity is critical.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 CritMultiplier = 1;
+
+    /// <summary>
+    ///     Damage multiplier if the devoured entity is dead.
+    /// </summary>
+    [DataField]
+    public FixedPoint2 DeadMultiplier = 0.25;
```
Comment (AJCM-git) [inline Content.Shared/Devour/Components/DevouredComponent.cs:53]: Should prob just apply the same amount of damage while you are alive and only stop once dead to make the code and mechanics more straight-forward

File: Content.Shared/Devour/Components/DevouredComponent.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    ///     Stores the original MobStates the entity was allowed to be healed in.
+    ///     This is so it can be returned to it's original state when the component is removed.
+    /// </summary>
+    [DataField]
+    public List<MobState>? OriginalAllowedMobStates;
+
+    /// <summary>
+    ///     Stores the stomach damage so it can be deducted from the devoured entity when the damage is altered.
+    /// </summary>
+    public DamageSpecifier? StomachDamage;
```
Comment (AJCM-git) [inline Content.Shared/Devour/Components/DevouredComponent.cs:22]: datafield?

File: Content.Shared/Devour/Components/DevourerComponent.cs
Comment (AJCM-git) [inline Content.Shared/Devour/Components/DevourerComponent.cs:1]: VV.RW is now inferred so you can put only datafield

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Server/Devour/DevourSystem.cs (hunk 13 above)
Comment (Kadeo64) [reply Content.Server/Devour/DevourSystem.cs:46]: i'm like 90% sure the dragon doesn't have a coded limit but if you devour the integer limit's worth of bodies it'll probably cause a server crash

Comment (AJCM-git) [reply Content.Server/Devour/DevourSystem.cs:46]: then either there needs to be a cap in the stomach or in the healing factor

**Comment by AJCM-git** [comment]
@Arendian you coming back to this?

**Comment by deltanedas** [comment]
calling dibs on this

**Comment by deltanedas** [comment]
this ded
